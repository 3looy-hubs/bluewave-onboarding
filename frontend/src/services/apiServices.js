import axios from 'axios';
import { API_BASE_URL } from '../utils/constants';

// Create an instance of axios
const apiClient = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});


// Function to handle login
export const login = async (email, password) => {
  try {
    const response = await apiClient.post('/login', { email, password });
    return response.data;
  } catch (error) {
    console.error('Login error:', error.response);
    throw error;
  }
};

// Function to handle sign up
export const signUp = async (userData) => {
  try {
    const response = await apiClient.post('/signup', userData);
    return response.data;
  } catch (error) {
    console.error('Sign up error:', error.response);
    throw error;
  }
};

export default apiClient;



