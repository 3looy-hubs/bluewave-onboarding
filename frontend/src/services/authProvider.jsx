import React, { useState, useEffect, useContext } from 'react';
import {apiClient} from './apiClient';

const AuthContext = React.createContext();

export const useAuth = () => {
    return useContext(AuthContext);
};

export const AuthProvider = ({ children }) => {
    const [isLoggedIn, setIsLoggedIn] = useState(false);

    useEffect(() => {
        const fetchUser = async () => {
            try {
                const authToken = localStorage.getItem('authToken');
                if (!authToken) {
                    setIsLoggedIn(false);
                    return;
                }
                const response = await apiClient.get('/users/current-user');
                if (response.status === 200 && response.data.user) {
                    setIsLoggedIn(true);
                }
                else{
                    setIsLoggedIn(false);
                }
            } catch (error) {
                localStorage.removeItem('authToken');
            }      
        };
        fetchUser();
    }, []);

    const loginAuth = () => {
        setIsLoggedIn(true);
    };

    const logoutAuth = () => {
        setIsLoggedIn(false);
    };

    return (
        <AuthContext.Provider value={{isLoggedIn, loginAuth, logoutAuth}}>
            {children}
        </AuthContext.Provider>
    );
};
