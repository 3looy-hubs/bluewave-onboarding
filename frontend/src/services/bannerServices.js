import {apiClient} from './apiClient'; 

export const addBanner = async (bannerData) => {
  try {
    const response = await apiClient.post('/banner/add_banner', bannerData);
    console.log(response)
    return response.data;
  } catch (error) {
    console.error('Add Banner error:', error);
    throw error;
  }
};
