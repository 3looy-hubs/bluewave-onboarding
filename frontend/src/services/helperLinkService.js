import { apiClient } from "./apiClient";

export const getHelpers = async () => {
  try {
    const response = await apiClient.get(`/helper-link`);
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Update helper link error:", error.response);
    throw error;
  }
};
export const getHelperById = async (id) => {
  try {
    const response = await apiClient.get(`/helper-link/${id}`);
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Update helper link error:", error.response);
    throw error;
  }
};

export const createHelper = async (helper) => {
  try {
    const response = await apiClient.post(`/helper-link`, {
      title: helper.title,
    });
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Update helper link error:", error.response);
    throw error;
  }
};

export const updateHelper = async (helper) => {
  try {
    const response = await apiClient.put(`/helper-link/${helper.id}`, {
      title: helper.title,
      headerBackgroundColor: helper.headerBackgroundColor,
      linkFontColor: helper.linkFontColor,
      iconColor: helper.iconColor,
      userId: helper.userId,
    });
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Update helper link error:", error.response);
    throw error;
  }
};

export const deleteHelper = async (id) => {
  try {
    const response = await apiClient.delete(`/helper-link/${id}`);
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Delete helper link error:", error.response);
    throw error;
  }
};
