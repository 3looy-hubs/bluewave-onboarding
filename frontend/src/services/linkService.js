import { apiClient } from "./apiClient";

export const createLink = async (link) => {
  try {
    const response = await apiClient.post(`/link`, {
      title: link.title,
      url: link.url,
      order: link.order,
      target: link.target ? "_blank" :  "_self"
    });
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Update Link error:", error.response);
    throw error;
  }
}

export const getLinks = async (helperId) => {
  try {
    const response = await apiClient.get(`/link?helperId=${helperId}`);
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Update Link error:", error.response);
    throw error;
  }
};

export const updateLink = async (link) => {
  try {
    const response = await apiClient.put(`/link/${link.id}`, {
      title: link.title,
      url: link.url,
      order: link.order,
      target: link.target ? "_blank" :  "_self"
    });
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Update Link error:", error.response);
    throw error;
  }
};

export const deleteLink = async (id) => {
  try {
    const response = await apiClient.delete(`/link/${id}`);
    if (response.status >= 400) throw new Error(response.data);
    return response.data;
  } catch (error) {
    console.error("Delete Link error:", error.response);
    throw error;
  }
};
