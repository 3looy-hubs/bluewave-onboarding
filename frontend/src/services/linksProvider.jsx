import { createContext, useMemo, useState } from "react";

export const HelperLinkContext = createContext({});

const HelperLinkProvider = ({ children }) => {
  const [helper, setHelper] = useState({});
  const [links, setLinks] = useState([]);
  const [showSettings, setShowSettings] = useState(false);
  const [itemToDelete, setItemToDelete] = useState(null);
  const [isPopupOpen, setIsPopupOpen] = useState(false);
  const [deletedLinks, setDeletedLinks] = useState([]);
  const [linkToEdit, setLinkToEdit] = useState(null);

  const toggleSettings = (e, link = null) => {
    if (e.target.closest("#delete") || e.target.closest("#drag")) return;
    if (showSettings) {
      renderLinks();
    }
    if (!showSettings && link) {
      localStorage.setItem("newLink", JSON.stringify(link));
    }
    setShowSettings(!showSettings); // Toggle the settings visibility
  };

  const value = useMemo(
    () => ({
      helper,
      setHelper,
      links,
      setLinks,
      showSettings,
      setShowSettings,
      toggleSettings,
      itemToDelete,
      setItemToDelete,
      isPopupOpen,
      setIsPopupOpen,
      deletedLinks,
      setDeletedLinks,
      linkToEdit,
      setLinkToEdit,
    }),
    [helper, links, showSettings, isPopupOpen, linkToEdit]
  );
  return (
    <HelperLinkContext.Provider value={value}>
      {children}
    </HelperLinkContext.Provider>
  );
};

export default HelperLinkProvider;
