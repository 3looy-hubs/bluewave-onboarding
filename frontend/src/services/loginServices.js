import {apiClient} from './apiClient';  // Import the existing apiClient
import axios from 'axios';

// Create a new instance of apiClient with 'auth' appended to the baseURL
const authClient = axios.create({
  ...apiClient.defaults,
  baseURL: `${apiClient.defaults.baseURL}auth`,
});

// Function to handle login
export const login = async (email, password) => {
  try {
    const response = await authClient.post('/login', { email, password });
    const token = response.data.token;
    localStorage.setItem('authToken', token);

    return response.data;
  } catch (error) {
    console.error('Login error:', error.response);
    throw error;
  }
};

// Function to handle logout
export const logout = async () => {
  try {
    await apiClient.post('auth/logout');
    localStorage.removeItem('authToken');
    return true;
  } catch (error) {
    console.error('Logout error:', error.response);
    localStorage.removeItem('authToken');
    throw error;
  }
};

// Function to handle sign up
export const signUp = async (userData) => {
  try {
    const response = await authClient.post('/register', userData);
    const token = response.data.token;
    localStorage.setItem('authToken', token);
    return response.data;
  } catch (error) {
    console.error('Sign up error:', error.response);
    throw error;
  }
};

// Function to handle sign up
export const forgotPassword = async (userData) => {
  try {
    const response = await authClient.post('/forget-password', userData);
    return response.data;
  } catch (error) {
    console.error(error.response);
    throw error;
  }
};

export const resetPassword = async (userData) => {
  try {
    const response = await authClient.post('/reset-password', userData);
    return response.data;
  } catch (error) {
    console.error(error.response);
    throw error;
  }
};

export default authClient;



