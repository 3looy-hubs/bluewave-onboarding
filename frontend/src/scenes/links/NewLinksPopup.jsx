import PropTypes from "prop-types";
import React, { useContext, useEffect, useState } from "react";
import Settings from "../../components/Links/Settings/Settings";
import Preview from "../../products/LinkPreview";
import { createHelper, updateHelper } from "../../services/helperLinkService";
import { deleteLink, getLinkById } from "../../services/linkService";
import { HelperLinkContext } from "../../services/linksProvider";
import GuideTemplate from "../../templates/GuideTemplate/GuideTemplate";
import { emitToastError } from "../../utils/guideHelper";
import toastEmitter, { TOAST_EMITTER_KEY } from "../../utils/toastEmitter";
import LinkAppearance from "./LinkPageComponents/LinkAppearance";
import LinkContent from "./LinkPageComponents/LinkContent";
import { useDialog } from "../../templates/GuideTemplate/GuideTemplateContext";
import s from "./LinkPage.module.scss";
import { getHelperById } from "../../services/helperLinkService";

const NewLinksPopup = ({
  autoOpen = false,
  isEdit,
  setIsEdit,
  itemId,
  setItemsUpdated
}) => {
  const [activeBtn, setActiveBtn] = useState(0);
  const [currentHelper, setCurrentHelper] = useState({});
  const [currentLinks, setCurrentLinks] = useState([]);

  const {
    showSettings,
    helper,
    setHelper,
    links,
    deletedLinks,
    setLinks,
    helperToEdit,
    setHelperToEdit,
  } = useContext(HelperLinkContext);

  const { openDialog, closeDialog, isOpen } = useDialog();
  const fetchHelperData = async () => {
    const { links, ...data } = await getHelperById(id);
    setCurrentHelper(data);
    setCurrentLinks(links.sort((a, b) => a.order - b.order));
    setHelperToEdit(itemId);
  }

  useEffect(() => {
    if (autoOpen) {
      openDialog();
    }
  }, [autoOpen, openDialog]);

  useEffect(() => {
    handleLinks
    setHelper(currentHelper);
    if (currentLinks?.length) {
      setLinks(currentLinks);
    }
    if (isEdit) {
      fetchHelperData()
    }
    else {
      setCurrentHelper({
        title: "",
        headerBackgroundColor: "#F8F9F8",
        linkFontColor: "#344054",
        iconColor: "#7F56D9",
      });
      setCurrentLinks([]);
    }
    if (!isOpen) {
      setLinks([]);
      setHelper({});
      setHelperToEdit(null);
    }
  }, [openDialog, isOpen]);

  const buildToastError = (msg) =>
    msg.response
      ? msg
      : {
        response: { data: { errors: [{ msg }] } },
      };

  const handleLinks = async (item) => {
    const { id, ...link } = item;
    if (!link?.title.trim() || !link?.url.trim()) {
      emitToastError(buildToastError("Title and URL are required"));
      return null;
    }
    try {
      const exists = await getLinkById(id);
      if (exists?.id) return { ...link, id };
      return { ...link };
    } catch (err) {
      emitToastError(err);
      return null;
    }
  };

  const handleSaveHelper = async () => {
    let newHelper;
    const formattedLinks = await Promise.all(links.map(handleLinks));
    try {
      if (formattedLinks.some((it) => !it)) {
        return null;
      }
      newHelper = await (helperToEdit
        ? updateHelper(helper, formattedLinks)
        : createHelper(helper, formattedLinks));
      setHelper(newHelper);
      setItemsUpdated((prevState) => !prevState);
      closeDialog();
    } catch (err) {
      emitToastError(buildToastError(err));
      return null;
    }
    if (helperToEdit && deletedLinks.length) {
      await Promise.all(
        deletedLinks.map(async (it) => {
          try {
            return await deleteLink({ ...it, helperId: helperToEdit });
          } catch (err) {
            emitToastError(err);
            return null;
          }
        })
      );
    }
    if (newHelper) {
      const toastMessage = helper.id
        ? "You edited this Helper Link"
        : "New Helper Link saved";
      toastEmitter.emit(TOAST_EMITTER_KEY, toastMessage);
      setShowNewLinksPopup(false);
      setHelper({});
      setLinks([]);
      setHelperToEdit(null);
    }
  };

  const rightContent = () => <Preview />;
  const leftContent = () => <LinkContent />;
  const leftAppearance = () => <LinkAppearance />;



  return (
    <div className={s.new__container}>
      <GuideTemplate
        title={isEdit ? "Edit Helper Link" : "New Helper Link"}
        activeButton={activeBtn}
        handleButtonClick={setActiveBtn}
        rightContent={rightContent}
        leftContent={leftContent}
        leftAppearance={leftAppearance}
        onSave={handleSaveHelper}
      />
      {showSettings && <Settings />}
    </div>
  );
};

NewLinksPopup.propTypes = {
  currentHelper: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    headerBackgroundColor: PropTypes.string,
    linkFontColor: PropTypes.string,
    iconColor: PropTypes.string,
  }),
  currentLinks: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      title: PropTypes.string,
      url: PropTypes.string,
      target: PropTypes.bool,
      x: PropTypes.number,
      y: PropTypes.number,
    })
  ),
  setShowNewLinksPopup: PropTypes.func,
  helperState: PropTypes.shape({
    isEdit: PropTypes.bool,
    id: PropTypes.number,
  }),
};

export default NewLinksPopup;