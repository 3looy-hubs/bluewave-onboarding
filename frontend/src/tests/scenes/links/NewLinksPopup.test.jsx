import { render } from "@testing-library/react";
import { act } from "react";
import { BrowserRouter as Router } from "react-router-dom";
import { describe, expect, it, vi } from "vitest";
import NewLinksPopup from "../../../scenes/links/NewLinksPopup";
import { AuthProvider } from "../../../services/authProvider";
import HelperLinkProvider from "../../../services/linksProvider";
import * as loginServices from "../../../services/loginServices";

vi.mock("react-router-dom", async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    useNavigate: () => vi.fn(), // Mock useNavigate to avoid actual routing
  };
});

vi.mock("../../../services/loginServices");
vi.mock("../../../services/helperLinkService");
vi.mock("../../../services/linkService");

const logIn = async () => {
  vi.spyOn(loginServices, "signUp").mockResolvedValueOnce({
    token: "mockAuthToken",
  });

  await act(async () => {
    await loginServices.signUp({
      name: "test",
      surname: "user",
      email: "test_user@example.com",
      password: "password123",
    });
    localStorage.setItem("authToken", "mockAuthToken");
  });
};

const renderPopup = () => {
  return render(
    <Router>
      <AuthProvider>
        <HelperLinkProvider>
          <NewLinksPopup
            currentHelper={{
              title: "",
              headerBackgroundColor: "#F8F9F8",
              linkFontColor: "#344054",
              iconColor: "#7F56D9",
            }}
          />
        </HelperLinkProvider>
      </AuthProvider>
    </Router>
  );
};

describe("Test Helper Link popup", () => {
  it("Tests if the popup is rendered correctly", async () => {
    await logIn();
    const { asFragment } = renderPopup();
    expect(asFragment).to.matchSnapshot();
  });
  it.todo("should open the settings drawer when add new link is clicked");
  it.todo(
    "should add thee new link to links list and to the preview when a new link is created"
  );
  it.todo("should change the order of the links on preview after drag");
  it.todo("should change the title on the preview when a title is typed");
  it.todo(
    "should change the background color of the header if the color is changed"
  );
  it.todo("should change the link text color if the color is changed");
  it.todo("should change the icon color if the color is changed");
});
