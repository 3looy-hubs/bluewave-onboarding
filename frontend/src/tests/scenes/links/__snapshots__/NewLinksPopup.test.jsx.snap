// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test Helper Link popup > Tests if the popup is rendered correctly 1`] = `[Function]`;
